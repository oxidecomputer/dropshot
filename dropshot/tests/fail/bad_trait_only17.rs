// Copyright 2024 Oxide Computer Company

#![allow(unused_imports)]

// Test for a trait with a macro invocation inside that has an endpoint
// annotation. (Endpoint annotations can only live on functions.)

macro_rules! life_the_universe_and_everything {
    () => {
        fn the_answer() -> u32;
    };
}

#[dropshot::api_description]
trait MyApi {
    type Context;

    #[endpoint { method = GET }]
    life_the_universe_and_everything!();
}

enum MyImpl {}

// This should not produce errors about the trait or any of the items within
// being missing.
impl MyApi for MyImpl {
    type Context = ();

    fn the_answer() -> u32 {
        42
    }
}

fn main() {
    // These items should be generated and accessible.
    my_api::api_description::<MyImpl>();
    my_api::stub_api_description();
}
