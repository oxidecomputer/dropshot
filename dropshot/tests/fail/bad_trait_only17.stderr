error: macro `life_the_universe_and_everything` marked as endpoint is not a method
  --> tests/fail/bad_trait_only17.rs:18:5
   |
18 |     #[endpoint { method = GET }]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
